lib/run/frame.ts

```diff
@@ -17,83 +18,182 @@ export interface FrameOptions<T> {
 }
 
 export function createFrame<T>(options: FrameOptions<T>): Frame<T> {
-  let { operation, parent } = options;
-  let children = new Set<Frame>();
-  let block = createBlock(operation);
-  let context = Object.create(parent?.context ?? {});
-  let results = createEventStream<void, Result<void>>();
+  return evaluate<Frame<T>>(function* () {
+    let { operation, parent } = options;
+    let children = new Set<Frame>();
+    let context = Object.create(parent?.context ?? {});
+    let thunks: IteratorResult<Thunk, Result<T>>[] = [{
+      done: false,
+      value: $next(void 0),
+    }];
+
+    let crash: Error | undefined = void 0;
+
+    let interrupt = () => {};
+
+    let abort = (reason?: Error) => {
+      if (!frame.aborted) {
+        frame.aborted = true;
+        crash = reason;
+        thunks.unshift({ done: false, value: $abort() });
+        interrupt();
+      }
+    };
+
+    let [setResults, results] = yield* createValue<FrameResult<T>>();
+
+    let frame = yield* shiftSync<Frame<T>>((k) => {
+      let self: Frame<T> = create<Frame<T>>("Frame", { id: ids++, context }, {
+        createChild<X>(operation: () => Operation<X>) {
+          let child = createFrame<X>({ operation, parent: frame });
+          children.add(child);
+          evaluate(function* () {
+            yield* child;
+            children.delete(child);
+          });
+          return child;
+        },
+        enter: () => {
+          let task = createTask(self);
+          self.enter = () => task;
+          k.tail(self);
```

Comment:
This is a red flag to me and something I find in general very confusing to reason about.  We are passing `self` to `createTask` but we are still defining the shape of `self` at this point.

If `createTask` needs the frame it seems like `createTask` could be merged into `createFrame` instead of it being passed in, especially because the task needs an event for when the frame is completed/halted.  Thoughts?

Reply:
I agree. The main reason to keep `createTask()`  separate was to make unrolling the `Future` logic tractable. Now that this is done, it makes sense to attempt yet another level of unrolling.

Reply:
I tried unrolling it without much success. I don't think we're still determining the shape of `Frame` at this point. This was part of the work to make the operations and promises and all of the moving bits of `task` lazy so that they never get created unless you are using them explicitly. I disentangled the `enter()` method from the `getTask()`. This means that now you can enter a frame without actually creating a task _at all_. Which I think is much better.

Finally, for the reference to `self`. It should now be written in a way where we could use `this` and `class` syntax if we wanted to.

Let me know what you think.